vars-on-top conflicts with `export`

<!--
This template is for bug reports. If you are reporting a bug, please continue on. If you are here for another reason, please see below:

1. To propose a new rule: http://eslint.org/docs/developer-guide/contributing/new-rules
2. To request a change: http://eslint.org/docs/developer-guide/contributing/changes
3. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**What version of ESLint are you using?**
2.5.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```js
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "env": {
    "node": true,
    "es6": true
  },
  "rules": {
    "strict": [2, "global"],

    "no-var": 0,
    "prefer-const": 0,

    // variables
    "no-shadow": [2, {
      "builtinGlobals": true
    }],
    "no-shadow-restricted-names": 2,
    "no-unused-vars": [2, {
      "vars": "all",
      "args": "after-used"
    }],
    "no-use-before-define": [2, {
      "functions": false
    }],

    // expressions
    "no-unused-expressions": 2,

    // possible errors
    "comma-dangle": [2, "always-multiline"],
    "no-cond-assign": [2, "except-parens"],
    "no-console": [2, {
      "allow": ["error", "warn", "info", "dir"]
    }],
    "no-debugger": 2,
    "no-alert": 1,
    "no-constant-condition": 2,
    "no-dupe-keys": 2,
    "no-duplicate-case": 2,
    "no-empty": 2,
    "no-ex-assign": 2,
    "no-extra-boolean-cast": 2,
    "no-extra-semi": 2,
    "no-func-assign": 2,
    "no-inner-declarations": 2,
    "no-invalid-regexp": 2,
    "no-irregular-whitespace": 2,
    "no-obj-calls": 2,
    "no-reserved-keys": 0,
    "no-sparse-arrays": 2,
    "no-unreachable": 2,
    "use-isnan": 2,
    "block-scoped-var": 2,

    // best practices
    "consistent-return": 2,
    "curly": [2, "all"],
    "default-case": 0,
    "dot-notation": [2, {
      "allowKeywords": true
    }],
    "eqeqeq": 0,
    "guard-for-in": 2,
    "no-caller": 2,
    "no-else-return": 2,
    "no-eq-null": 0,
    "no-eval": 2,
    "no-extend-native": 2,
    "no-extra-bind": 2,
    "no-fallthrough": 2,
    "no-floating-decimal": 2,
    "no-implied-eval": 2,
    "no-lone-blocks": 2,
    "no-loop-func": 2,
    "no-multi-str": 2,
    "no-native-reassign": 2,
    "no-new": 2,
    "no-new-func": 2,
    "no-new-wrappers": 2,
    "no-octal": 2,
    "no-octal-escape": 2,
    "no-param-reassign": 2,
    "no-proto": 2,
    "no-redeclare": 2,
    "no-return-assign": [2, "except-parens"],
    "no-script-url": 2,
    "no-self-compare": 2,
    "no-sequences": 2,
    "no-throw-literal": 2,
    "no-with": 2,
    "radix": 2,
    "vars-on-top": 2,
    "wrap-iife": [2, "inside"],
    "yoda": 2,

    // style
    "indent": [2, 2],
    "brace-style": [2, "1tbs", {
      "allowSingleLine": true
    }],
    "quotes": [2, "single"],
    "camelcase": [2, {
    "properties": "always"
    }],
    "comma-spacing": [2, {
      "before": false,
      "after": true
    }],
    "comma-style": [2, "last"],
    "eol-last": 2,
    "func-names": 2,
    "key-spacing": [2, {
        "beforeColon": false,
        "afterColon": true
    }],
    "new-cap": [2, {
      "newIsCap": true
    }],
    "no-multiple-empty-lines": [2, {
      "max": 1
    }],
    "no-nested-ternary": 2,
    "no-new-object": 2,
    "no-spaced-func": 2,
    "no-trailing-spaces": 2,
    "no-extra-parens": [2, "all", {
      "conditionalAssign": false
    }],
    "no-underscore-dangle": 0,
    "one-var": [2, "never"],
    "padded-blocks": [2, "never"],
    "semi": [2, "always"],
    "semi-spacing": [2, {
      "before": false,
      "after": true
    }],
    "keyword-spacing": [2, {
      "before": true,
      "after": true
    }],
    "space-before-blocks": 2,
    "space-before-function-paren": [2, "never"],
    "space-infix-ops": 2,
    "spaced-comment": [2, "always"],
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```js
// test.js
export var a;
var b;
```

**What did you expect to happen?**
I wouldn't expect the `vars-on-top` rule to error because even though one of the variables is exported, they are still all defined at the top of the file.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
/home/olmo/projects/atom/tree-view-git-branch/test.js
  1:8  error  All 'var' declarations must be at the top of the function scope  vars-on-top
  2:1  error  All 'var' declarations must be at the top of the function scope  vars-on-top
  2:5  error  'b' is defined but never used                                    no-unused-vars

✖ 3 problems (3 errors, 0 warnings)
```

--- a/lib/rules/vars-on-top.js
+++ b/lib/rules/vars-on-top.js
@@ -55,7 +55,9 @@ module.exports = function(context) {
         }
 
         for (; i < l; ++i) {
-            if (statements[i].type !== "VariableDeclaration") {
+            if (statements[i].type !== "VariableDeclaration" &&
+                    (statements[i].type !== "ExportNamedDeclaration" ||
+                    statements[i].declaration.type !== "VariableDeclaration")) {
                 return false;
             }
             if (statements[i] === node) {
@@ -104,6 +106,12 @@ module.exports = function(context) {
             var grandParent = ancestors.pop();
 
             if (node.kind === "var") { // check variable is `var` type and not `let` or `const`
+                if (parent.type === "ExportNamedDeclaration") {
+                    node = parent;
+                    parent = grandParent;
+                    grandParent = ancestors.pop();
+                }
+
                 if (parent.type === "Program") { // That means its a global variable
                     globalVarCheck(node, parent);
                 } else {

--- a/tests/lib/rules/vars-on-top.js
+++ b/tests/lib/rules/vars-on-top.js
@@ -159,7 +159,40 @@ ruleTester.run("vars-on-top", rule, {
         {code: "import { square, diag } from 'lib'; 'use strict'; var y; function f() { 'use strict'; var x; var y; f(); }", parserOptions: { sourceType: "module" }},
         {code: "import { default as foo } from 'lib'; 'use strict'; var y; function f() { 'use strict'; var x; var y; f(); }", parserOptions: { sourceType: "module" }},
         {code: "import 'src/mylib'; 'use strict'; var y; function f() { 'use strict'; var x; var y; f(); }", parserOptions: { sourceType: "module" }},
-        {code: "import theDefault, { named1, named2 } from 'src/mylib'; 'use strict'; var y; function f() { 'use strict'; var x; var y; f(); }", parserOptions: { sourceType: "module" }}
+        {code: "import theDefault, { named1, named2 } from 'src/mylib'; 'use strict'; var y; function f() { 'use strict'; var x; var y; f(); }", parserOptions: { sourceType: "module" }},
+        {
+            code: [
+                "export var x;",
+                "var y;",
+                "var z;"
+            ].join("\n"),
+            parserOptions: {
+                ecmaVersion: 6,
+                sourceType: "module"
+            }
+        },
+        {
+            code: [
+                "var x;",
+                "export var y;",
+                "var z;"
+            ].join("\n"),
+            parserOptions: {
+                ecmaVersion: 6,
+                sourceType: "module"
+            }
+        },
+        {
+            code: [
+                "var x;",
+                "var y;",
+                "export var z;"
+            ].join("\n"),
+            parserOptions: {
+                ecmaVersion: 6,
+                sourceType: "module"
+            }
+        }
     ],
 
     invalid: [
@@ -402,6 +435,29 @@ ruleTester.run("vars-on-top", rule, {
         },
         {
             code: "function f() { 'use strict'; var x; 'directive';  var y; f(); }",
+            errors: [{message: "All 'var' declarations must be at the top of the function scope.", type: "VariableDeclaration"}]
+        },
+        {
+            code: [
+                "export function f() {}",
+                "var x;"
+            ].join("\n"),
+            parserOptions: {
+                ecmaVersion: 6,
+                sourceType: "module"
+            },
+            errors: [{message: "All 'var' declarations must be at the top of the function scope.", type: "VariableDeclaration"}]
+        },
+        {
+            code: [
+                "var x;",
+                "export function f() {}",
+                "var y;"
+            ].join("\n"),
+            parserOptions: {
+                ecmaVersion: 6,
+                sourceType: "module"
+            },
             errors: [{message: "All 'var' declarations must be at the top of the function scope.", type: "VariableDeclaration"}]
         }
     ]

